OIDC: Limit number of dynamic tenants

snazy opened the pull request:
Introduce a new configuration option `quarkus.oidc.dynamic-tenant-limit` as an opt-in to limit the number of active dynamic OIDC tenants. The default is no dynamic tenant limit.

This change adds the field `volatile long TenantConfigContext.lastUsed`, and removes the field `TenantConfigContext.ready` to keep the heap footprint the same. The `lastUsed` field is initially set to "now" and updated when the dynamic tenant is being accessed.

When a new dynamic tenant is about to be added to the dynamic tenants map, eviction runs, if there are more dynamic tenants than configured via `dynami-tenant-limit`. The eviction algo is built to iterate over the dynamic tenants only once - it may need to iterate more often, if one of the eviction candidates has been accessed in the mean time. There is no linked-list structure to form an LRU list/queue, as that likely causes more runtime overhead (pointer updates, synchronization) than the cost of iterating over the list once in a while. The assumption is that the map of dynamic tenants has not a lot of "churn".

Fixes #42803

sberyozkin commented on extensions/oidc/runtime/pom.xml:
```diff
@@ -45,6 +45,10 @@
             <groupId>jakarta.annotation</groupId>
             <artifactId>jakarta.annotation-api</artifactId>
         </dependency>
+        <dependency>
```
I'd like to avoid it, please see my comments in the issue

snazy commented:
Marking this PR as draft, I'm reworking the changes. The current state is rather a preparation.

sberyozkin commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/TenantConfigBean.java:
```diff
@@ -1,66 +1,115 @@
+        }
+    }
+
+    public TenantConfigContext getExistingTenantConfigContext(String tenantId) {
```
I'd consider `Dynamic` instead of `Existing`

sberyozkin commented:
Hi @snazy Thanks, I've had a quick look, and it looks good

snazy commented:
> Hi @snazy Thanks, I've had a quick look, and it looks good

Okay, factored out the preparation into #43007

snazy committed:
OIDC: Encapsulate static/dynamic tenants maps in `TenantConfigBean`

Lets `TenantConfigBean` be the sole "owner" of the static/dynamic tenants maps, adds/changes accessor methods for tenants. Also introduces a functional interface to create tenants.

No functional change, only moving code around.

Also some minor code cleanups (reducing the amount of warnings in IntelliJ).

snazy committed:
OIDC: Don't let static tenants "escape" into the dynamic tenants map

snazy committed:
OIDC: Limit number of dynamic tenants

Introduce a new configuration option `quarkus.oidc.dynamic-tenant-limit` as an opt-in to limit the number of active dynamic OIDC tenants. The default is no dynamic tenant limit.

This change adds the field `volatile long TenantConfigContext.lastUsed`, and removes the field `TenantConfigContext.ready` to keep the heap footprint the same. The `lastUsed` field is initially set to "now" and updated when the dynamic tenant is being accessed.

When a new dynamic tenant is about to be added to the dynamic tenants map, eviction runs, if there are more dynamic tenants than configured via `dynami-tenant-limit`. The eviction algo is built to iterate over the dynamic tenants only once - it may need to iterate more often, if one of the eviction candidates has been accessed in the mean time. There is no linked-list structure to form an LRU list/queue, as that likely causes more runtime overhead (pointer updates, synchronization) than the cost of iterating over the list once in a while. The assumption is that the map of dynamic tenants has not a lot of "churn".

snazy commented:
Here's an approach that limits the number of active dynamic tenants, if enabled. Updated the PR description accordingly.

Still in draft, because this PR depends on #43007 and #43110

snazy committed:
"Sharpen" Java visibility of `TenantConfigBean` members + re-add removed public functions, simplify some more code

sberyozkin commented:
Thanks @snazy, I'll have a closer look once the PRs it depends on are merged.
One thing I forgot to mention, is that we have an in memory abstract cache implemenation which is used in 4 cases already.

For, look at the back channel logout cache and corresponding configuration properties in the Logout config group.
IMHO the same cache should be used here

Now that you have encapsulated the dynamic map nicely, I believe we should just have DynamicTenantsMap extending the abstract cache impl, instead of the dynamic tenants map, which should also terminate the optional Vertx timer on the shutdown

radcortez commented:
Stale PR. Please reopen it if it still makes sense.
